### PR TITLE
Scaffold Sempai Core and Facade with core types, diagnostics, tests

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Scaffold the Sempai core infrastructure by introducing two new crates (sempai-core and sempai), wiring them into the workspace, and laying the groundwork for a stable public API and future backend integration.

## Changes
- Add two new crates and wire them into the workspace:
  - crate: sempai-core
  - crate: sempai
  - Update workspace members to include both crates.
- Core data model (sempai-core):
  - Language enum (Rust, Python, TypeScript, Go, Hcl) with serde support and non_exhaustive.
  - Span and LineCol types for source location tracking with serde derives.
  - CapturedNode and CaptureValue (Node, Nodes) for query bindings with adjacently tagged serde.
  - Match type holding rule_id, uri, span, optional focus, and captures.
  - EngineConfig with defaults and accessors; Default impl provided.
  - Diagnostic system: DiagnosticCode, SourceSpan, Diagnostic, DiagnosticReport with serialization and display implementations.
  - Diagnostic utilities: not_implemented feature constructor for common stub errors.
  - Tests: unit tests for language, span, capture, match, config, and diagnostics; behaviour tests for BDD style steps; feature files for core diagnostics and serialization.
- Facade engine (sempai):
  - QueryPlan struct with rule_id, language, and private placeholder plan field; accessors provided.
  - Engine struct with EngineConfig and stubbed methods:
    - compile_yaml(yaml) -> Err(DiagnosticReport::not_implemented("compile_yaml"))
    - compile_dsl(rule_id, language, dsl) -> Err(DiagnosticReport::not_implemented("compile_dsl"))
    - execute(plan, uri, source) -> Err(DiagnosticReport::not_implemented("execute"))
  - Re-exports in sempai::pub use to surface core types as stable public API.
- Documentation and planning artefacts:
  - docs/users-guide.md updated with Sempai query engine description and scope.
  - docs/roadmap.md updated to mark 4.1.1 as done.
  - docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md added with implementation plan summary.
- Testing and examples:
  - Unit tests for all new core types and engine façade behaviors.
  - Behaviour-driven test scaffolding for core and engine surface using rstest-bdd.
  - Feature files for sempai-core and sempai-engine scenarios added.

## Why this design
- Establishes a stable public API surface via sempai-core and the facade sempai, enabling progressive backend integration in subsequent tasks.
- Uses non_exhaustive enums and const fn constructors to improve extensibility and performance while keeping a strict public API surface.
- Serialization patterns (serde with adjacently tagged CaptureValue) chosen to accommodate tuple variants safely.
- Stubbed Engine methods provide clear, structured diagnostics (NotImplemented) without panicking or using unwraps, aligning with lint expectations.

## Testing plan
- Run cargo test for the workspace to validate core types, serialization round-trips, and diagnostic formatting.
- Run cargo doc -p sempai --no-deps with RUSTDOCFLAGS="-D warnings" to ensure documentation builds without warnings.
- Ensure unit tests cover:
  - Language display and serde round-trips
  - LineCol and Span serialization/deserialization
  - CapturedNode and CaptureValue serde behavior (including kind tag)
  - Match serialization and capture ordering via BTreeMap
  - EngineConfig defaults and custom construction
  - DiagnosticCode formatting and Diagnostic/DiagnosticReport behavior
- Validate BDD tests and feature files compile and the scenarios are wired to the step definitions.

## Notes
- All new APIs are designed to be extended in future iterations; existing public items are stable and ready for downstream usage.
- The engine methods are intentionally stubs returning NotImplemented diagnostics to satisfy lint rules and to provide a realistic surface area for downstream integration work.
- This PR introduces a substantial amount of new test scaffolding and docs to support the staged roadmap work.

📎 **Task**: https://www.devboxer.com/task/1868db31-b7d2-466a-8250-4d3148ddbe6b

## Summary by Sourcery

Introduce Sempai core and facade crates providing a stable query engine API surface, structured diagnostics, and engine configuration, and document their role in the workspace.

New Features:
- Add `sempai_core` crate defining core Sempai data model types for languages, spans, captures, matches, engine limits, and diagnostics.
- Add `sempai` facade crate that re-exports core types and exposes an `Engine` and `QueryPlan` entrypoint API for compiling and executing queries (currently stubbed).

Enhancements:
- Document the Sempai query engine architecture and crates in the user guide and repository layout, and mark the corresponding roadmap task as complete.

Documentation:
- Add an execution plan document for scaffolding `sempai_core` and `sempai`, detailing design, constraints, and validation steps.

Tests:
- Add unit and BDD-style tests for new core types, diagnostics, configuration, and the facade engine behaviour, including feature files for core and engine scenarios.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/sempai-core/src/tests/behaviour.rs](https://github.com/leynos/weaver/pull/68/files/b3f680ba9e22c3a66917ea8221d83e5a59053740#diff-c70512f1a0809854790b6bc8fb4d6415811eb1bbd0a8077e161b2289b7203e27)

Comment on file

~~~rust
fn given_span(
    world: &mut TestWorld,
    start_byte: u32,
    end_byte: u32,
    sl: u32,
    sc: u32,
    el: u32,
    ec: u32,
) {
    world.span = Some(Span::new(
        start_byte,
        end_byte,
        LineCol::new(sl, sc),
        LineCol::new(el, ec),
    ));
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=68&biomarker=Excess+Number+of+Function+Arguments&filename=crates%2Fsempai-core%2Fsrc%2Ftests%2Fbehaviour.rs&method=given_span)
given_span has 7 arguments, max arguments = 4

File: docs/execplans/5-1-1-entry-help.md [hunk 1]
```diff
@@ -188,12 +188,12 @@ satisfies roadmap task 5.1.1 in `../roadmap.md`.
...
   Fluent pipeline fails (via `NoOpLocalizer`), while genuine translations
   override it cleanly. The function composes the help block from individual
-  Fluent messages so translators can work on each line independently.
-  The English fallback values are centralised in a `bare_help` constants
-  module in `localizer.rs` so each string appears exactly once in the Rust
-  source; the `.ftl` file remains the canonical Fluent resource.  The
-  `fluent_and_fallback_outputs_are_identical` test guards against drift
-  between the two. Date/Author: 2026-02-26, updated 2026-02-28
+  Fluent messages so translators can work on each line independently. The
+  English fallback values are centralised in a `bare_help` constants module in
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/5-1-1-entry-help.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/sempai/src/tests/behaviour.rs](https://github.com/leynos/weaver/pull/68/files/bd5c5ed9b8564d17715a26137afee70381b1309d#diff-7e6ac3614493c2529c89000110987b32b2c85484bab969cd2a36721a2a1da351)

Comment on file

~~~rust
fn then_compilation_fails(world: &mut TestWorld, code: QuotedString) {
    let result = world
        .compile_result
        .as_ref()
        .expect("compile result should be set");
    let report = result.as_ref().expect_err("expected compilation failure");
    let first = report
        .diagnostics()
        .first()
        .expect("at least one diagnostic");
    let actual_code = format!("{}", first.code());
    assert_eq!(
        actual_code,
        code.as_str(),
        "expected code '{}', got '{actual_code}'",
        code.as_str()
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=68&biomarker=Code+Duplication&filename=crates%2Fsempai%2Fsrc%2Ftests%2Fbehaviour.rs&method=)
The module contains 2 functions with similar structure: then_compilation_fails,then_execution_fails

File: crates/sempai/src/engine.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        reason = "constructor will be used once compile_yaml and compile_dsl are implemented"
+    )]
+    #[must_use]
+    pub(crate) const fn new(rule_id: String, language: Language) -> Self {
+        Self {
+            rule_id,
+            language,
+            _plan: (),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai/src/engine.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/diagnostic.rs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+pub struct DiagnosticReport {
+    diagnostics: Vec<Diagnostic>,
+}
+
+impl DiagnosticReport {
+    /// Creates a report from a vector of diagnostics.
+    #[must_use]
+    pub const fn new(diagnostics: Vec<Diagnostic>) -> Self {
+        Self { diagnostics }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/diagnostic.rs:245]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/diagnostic.rs [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        span: Option<SourceSpan>,
+        notes: Vec<String>,
+    ) -> Self {
+        Self {
+            code,
+            message,
+            span,
+            notes,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/diagnostic.rs:181]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/diagnostic.rs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    /// Optional URI of the source file containing this span.
+    uri: Option<String>,
+}
+
+impl SourceSpan {
+    /// Creates a new source span.
+    #[must_use]
+    pub const fn new(start: u32, end: u32, uri: Option<String>) -> Self {
+        Self { start, end, uri }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/diagnostic.rs:114]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/capture.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    /// This is `None` when streaming mode omits text to avoid re-slicing.
+    pub text: Option<String>,
+}
+
+impl CapturedNode {
+    /// Creates a new captured node.
+    #[must_use]
+    pub const fn new(span: Span, kind: String, text: Option<String>) -> Self {
+        Self { span, kind, text }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/capture.rs:48]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/diagnostic.rs [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+///
+/// ```
+/// use sempai_core::DiagnosticCode;
+///
+/// let code = DiagnosticCode::ESempaiYamlParse;
+/// assert_eq!(format!("{code}"), "E_SEMPAI_YAML_PARSE");
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[non_exhaustive]
+pub enum DiagnosticCode {
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/diagnostic.rs:29]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/sempai-core/src/language.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+/// # Example
+///
+/// ```
+/// use sempai_core::Language;
+///
+/// let lang = Language::Rust;
+/// assert_eq!(format!("{lang}"), "rust");
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/language.rs:22]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The `QuotedString` parsing helpers and manual language string-to-enum mappings are duplicated across `sempai_core` and `sempai` BDD tests; consider extracting a small shared test utility (or adding a `FromStr` impl for `Language`) to reduce duplication and keep step definitions simpler.
- In `QueryPlan::new` you suppress `dead_code` via `#[expect]`; you could instead add a small unit test that constructs a `QueryPlan` via this constructor to exercise it and drop the lint suppression.
- The BDD steps in `crates/sempai-core/src/tests/behaviour.rs` manually parse range strings like `"10..42"` and `"2:0..4:0"`; extracting small parsing helpers (or using a more structured representation in the feature files) would make these tests less brittle and easier to extend.

## Individual Comments

### Comment 1
<location path="crates/sempai-core/src/tests/behaviour.rs" line_range="155-162" />
<code_context>
+// Then steps
+// ---------------------------------------------------------------------------
+
+#[then("the JSON contains key {key} with value {value}")]
+fn then_json_contains(world: &mut TestWorld, key: QuotedString, value: QuotedString) {
+    let json = world.json_output.as_ref().expect("JSON should be set");
+    let expected = format!("\"{}\":{}", key.as_str(), value.as_str());
+    assert!(
+        json.contains(&expected),
+        "expected JSON to contain '{expected}', got: {json}"
+    );
+}
+
</code_context>
<issue_to_address>
**suggestion:** JSON assertions are string-fragile; deserialize and assert on the structure instead

`then_json_contains` relies on `str::contains` with a handcrafted `"key":value` snippet, which is brittle to spacing/ordering changes or nested structures. Instead, deserialize `json_output` into `serde_json::Value` and assert on the parsed structure (e.g., `value[&*key]` or via `get`/`as_*`), so these checks stay robust as the JSON schema evolves.

Suggested implementation:

```rust
use std::str::FromStr;

use rstest::fixture;
use rstest_bdd_macros::{given, scenario, then, when};
use serde_json::Value;

use crate::{Diagnostic, DiagnosticCode, DiagnosticReport, Language, LineCol, Span};

```

```rust
+// Then steps
+// ---------------------------------------------------------------------------
+
+#[then("the JSON contains key {key} with value {value}")]
+fn then_json_contains(world: &mut TestWorld, key: QuotedString, value: QuotedString) {
+    let json = world.json_output.as_ref().expect("JSON should be set");
+
+    let parsed: Value =
+        serde_json::from_str(json).expect("JSON output should be valid JSON");
+
+    let key_str = key.as_str();
+    let actual = parsed
+        .get(key_str)
+        .unwrap_or_else(|| panic!("expected JSON to contain key `{key_str}`, got: {json}"));
+
+    // Interpret the expected value as JSON if possible (number, bool, object, etc.),
+    // otherwise fall back to a JSON string value.
+    let expected: Value = serde_json::from_str(value.as_str())
+        .unwrap_or_else(|_| Value::String(value.as_str().to_owned()));
+
+    assert_eq!(
+        actual, &expected,
+        "expected JSON key `{key_str}` to have value {expected:?}, got {actual:?}"
+    );
+}
+

```
</issue_to_address>

### Comment 2
<location path="crates/sempai/src/tests/engine_tests.rs" line_range="5" />
<code_context>
+    Span::new(0, 5, LineCol::new(0, 0), LineCol::new(0, 5))
+}
+
+#[test]
+fn captured_node_construction_and_accessors() {
+    let node = CapturedNode::new(
</code_context>
<issue_to_address>
**suggestion (testing):** Add a test that `Engine::execute` also returns a NOT_IMPLEMENTED diagnostic

We already verify `DiagnosticCode::NotImplemented` for `compile_yaml` and `compile_dsl`, but there’s no equivalent assertion for `Engine::execute`. Please add a test similar to `compile_yaml_returns_not_implemented` that calls `execute` and asserts both the `NotImplemented` code and that the diagnostic message references "execute".
</issue_to_address>

### Comment 3
<location path="crates/sempai-core/src/tests/capture_tests.rs" line_range="57-78" />
<code_context>
+    assert!(matches!(deserialized, CaptureValue::Node(_)));
+}
+
+#[test]
+fn capture_value_nodes_serde_round_trip() {
+    let nodes = vec![
+        CapturedNode::new(
+            sample_span(),
+            String::from("identifier"),
+            Some(String::from("a")),
+        ),
+        CapturedNode::new(
+            Span::new(10, 15, LineCol::new(1, 0), LineCol::new(1, 5)),
+            String::from("identifier"),
+            Some(String::from("b")),
+        ),
+    ];
+    let value = CaptureValue::Nodes(nodes);
+    let json = serde_json::to_string(&value).expect("serialize");
+
+    assert!(json.contains("\"kind\":\"nodes\""));
+
+    let deserialized: CaptureValue = serde_json::from_str(&json).expect("deserialize");
+    assert!(matches!(deserialized, CaptureValue::Nodes(ref v) if v.len() == 2));
+}
</code_context>
<issue_to_address>
**suggestion (testing):** Consider an explicit test for `CaptureValue::Nodes` with an empty node list

Right now we only verify `CaptureValue::Node` and `CaptureValue::Nodes` with multiple entries. Please also add a test where `CaptureValue::Nodes` is created from an empty `Vec<CapturedNode>` and round-tripped through serde, to document and lock in the behavior for ellipsis metavariables that bind zero nodes and prevent assumptions that `Nodes` is always non-empty.

```suggestion
#[test]
fn capture_value_nodes_serde_round_trip() {
    let nodes = vec![
        CapturedNode::new(
            sample_span(),
            String::from("identifier"),
            Some(String::from("a")),
        ),
        CapturedNode::new(
            Span::new(10, 15, LineCol::new(1, 0), LineCol::new(1, 5)),
            String::from("identifier"),
            Some(String::from("b")),
        ),
    ];
    let value = CaptureValue::Nodes(nodes);
    let json = serde_json::to_string(&value).expect("serialize");

    assert!(json.contains("\"kind\":\"nodes\""));

    let deserialized: CaptureValue = serde_json::from_str(&json).expect("deserialize");
    assert!(matches!(deserialized, CaptureValue::Nodes(ref v) if v.len() == 2));
}

#[test]
fn capture_value_empty_nodes_serde_round_trip() {
    let nodes: Vec<CapturedNode> = Vec::new();
    let value = CaptureValue::Nodes(nodes);
    let json = serde_json::to_string(&value).expect("serialize");

    assert!(json.contains("\"kind\":\"nodes\""));

    let deserialized: CaptureValue = serde_json::from_str(&json).expect("deserialize");
    assert!(matches!(deserialized, CaptureValue::Nodes(ref v) if v.is_empty()));
}
```
</issue_to_address>

### Comment 4
<location path="crates/sempai/src/engine.rs" line_range="37" />
<code_context>
+
+impl QueryPlan {
+    /// Creates a new query plan (crate-internal).
+    #[expect(
+        dead_code,
+        reason = "constructor will be used once compile_yaml and compile_dsl are implemented"
</code_context>
<issue_to_address>
**issue (review_instructions):** The `#[expect(dead_code, ...)]` reason suggests a temporary suppression but does not include the required `FIXME:` with a link.

Because this `dead_code` suppression is explicitly described as temporary (it will be used once `compile_yaml` and `compile_dsl` are implemented), it falls under the “fix is expected” case where the guidelines require `FIXME:` plus a link in the `reason`.

Consider changing the attribute to something like:

```rust
#[expect(
    dead_code,
    reason = "FIXME: link-to-issue – constructor will be used once compile_yaml and compile_dsl are implemented",
)]
```

and pointing the link to the relevant tracking issue or ticket.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Lints must not be silenced except as a **last resort**.

  * `#[allow]` is **forbidden**.
  * Only **narrowly scoped** `#[expect(lint, reason = "...")]` is allowed.
  * No lint groups, no blanket or file-wide suppression.
  * Include `FIXME:` with link if a fix is expected.

</details>
</issue_to_address>

### Comment 5
<location path="docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md" line_range="21" />
<code_context>
+
+Running `cargo doc -p sempai --no-deps` produces complete, warning-free public
+API documentation. Unit tests validate type construction, serde round-trips,
+and diagnostic formatting. BDD scenarios exercise the public API surface from a
+consumer's perspective.
+
</code_context>
<issue_to_address>
**suggestion (review_instructions):** "BDD" is used here without first expanding it (for example, "behaviour-driven development (BDD)").

To follow the acronym guidance, expand the term on its first occurrence in the document, for example: "behaviour-driven development (BDD) scenarios".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 6
<location path="docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md" line_range="60" />
<code_context>
+  (`docs/sempai-query-language-design.md` line 191).
+- Library crates use `thiserror`-derived error enums — no `eyre` or `anyhow`
+  (`AGENTS.md` lines 220-227).
+- All dependency versions use SemVer-compatible caret requirements
+  (`AGENTS.md` lines 206-216).
+- `rstest-bdd` v0.5.0 must be used for BDD tests
</code_context>
<issue_to_address>
**suggestion (review_instructions):** "SemVer" is introduced without definition (for example, "Semantic Versioning (SemVer)").

Please expand "SemVer" at its first occurrence, such as "Semantic Versioning (SemVer)-compatible caret requirements", to comply with the acronym rule.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: crates/sempai/src/tests/behaviour.rs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// ---------------------------------------------------------------------------
+// Scenario registration
+// ---------------------------------------------------------------------------
+
+#[scenario(path = "tests/features/sempai_engine.feature")]
+fn sempai_engine_behaviour(world: TestWorld) {
+    let _ = world;
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai/src/tests/behaviour.rs:130]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

File: crates/sempai-core/src/test_support.rs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+/// Parses by stripping surrounding double-quote characters.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct QuotedString(String);
+
+impl FromStr for QuotedString {
+    type Err = std::convert::Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(s.trim_matches('"').to_owned()))
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/test_support.rs:35]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

File: docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md [hunk 11]
```diff
@@ -0,0 +1,869 @@
...
+
+**F2.** Mark roadmap task 4.1.1 as done in `docs/roadmap.md` (line 347): change
+`- [ ] 4.1.1.` to `- [x] 4.1.1.`.
+
+**F3.** Record design decisions taken in the design document
+`docs/sempai-query-language-design.md` if any type definitions diverge from the
+design (e.g. serde tagging strategy for `CaptureValue`).
+
+**F4.** Run `make fmt` to format all changed files.
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

File: crates/sempai-core/src/test_support.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (leynos) [reply crates/sempai-core/src/test_support.rs:35]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

**Comment by leynos** [comment]
@sourcery-ai review

**Comment by leynos** [comment]
@coderabbitai have the following now been resolved?

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="crates/sempai-core/src/tests/match_tests.rs" line_range="51" />
<code_context>
+}
+
+#[test]
+fn match_serde_round_trip() {
+    let m = Match::new(
+        String::from("test-rule"),
</code_context>
<issue_to_address>
**suggestion (testing):** Add a serde round-trip test for Match with non-empty captures to verify nested CaptureValue serialization

The existing `match_serde_round_trip` test only covers a `Match` with an empty `BTreeMap` of captures. Please add another test that builds a `Match` with both `Node` and `Nodes` captures, round-trips it through JSON, and verifies that the capture keys and the inner `CaptureValue` variants (including their `kind` tag and payload) are preserved after deserialization.

Suggested implementation:

```rust
#[test]
fn match_serde_round_trip() {
    let m = Match::new(
        String::from("test-rule"),
        String::from("file:///test.py"),
        sample_span(),
        None,
        BTreeMap::new(),
    );
    let json = serde_json::to_string(&m).expect("serialize");
    let deserialized: Match = serde_json::from_str(&json).expect("deserialize");
    assert_eq!(deserialized.rule_id(), "test-rule");
    assert_eq!(deserialized.uri(), "file:///test.py");
}

#[test]
fn match_serde_round_trip_with_captures() {
    // Build a Match with both Node and Nodes captures so we can verify
    // that the CaptureValue enum (kind tag + payload) is preserved.
    let mut captures = BTreeMap::new();

    // These values are cloned so we can compare them after deserialization.
    let node_capture = CaptureValue::Node(sample_node());
    let nodes_capture = CaptureValue::Nodes(vec![sample_node(), sample_node()]);

    captures.insert(String::from("$node"), node_capture.clone());
    captures.insert(String::from("$nodes"), nodes_capture.clone());

    let m = Match::new(
        String::from("test-rule"),
        String::from("file:///test.py"),
        sample_span(),
        None,
        captures,
    );

    let json = serde_json::to_string(&m).expect("serialize with captures");
    let deserialized: Match = serde_json::from_str(&json).expect("deserialize with captures");

    assert_eq!(deserialized.rule_id(), "test-rule");
    assert_eq!(deserialized.uri(), "file:///test.py");

    let deserialized_captures = deserialized.captures();
    assert_eq!(deserialized_captures.len(), 2);
    assert!(deserialized_captures.contains_key("$node"));
    assert!(deserialized_captures.contains_key("$nodes"));

    // Verify that the CaptureValue variants and payloads survived the round-trip.
    let deserialized_node = deserialized_captures.get("$node").expect("node capture missing");
    let deserialized_nodes = deserialized_captures.get("$nodes").expect("nodes capture missing");

    match deserialized_node {
        CaptureValue::Node(_) => {}
        other => panic!("expected CaptureValue::Node for `$node`, got {:?}", other),
    }

    match deserialized_nodes {
        CaptureValue::Nodes(nodes) => {
            assert!(!nodes.is_empty(), "expected non-empty nodes payload for `$nodes`");
        }
        other => panic!("expected CaptureValue::Nodes for `$nodes`, got {:?}", other),
    }

    // If CaptureValue implements Eq, also assert exact payload equality.
    // This gives a stronger guarantee that the entire payload survives.
    #[allow(clippy::bool_assert_comparison)]
    {
        // Using a block so this compiles even if PartialEq/Eq are not implemented;
        // see additional_changes below.
        let _ = &node_capture;
        let _ = &nodes_capture;
    }
}

```

To make this compile and fully meet the intent, you may need to adjust a few details to match your actual types:

1. Ensure `CaptureValue` is in scope in this test module (e.g. `use crate::match::CaptureValue;` or similar) if it is not already imported.
2. Replace `sample_node()` with whatever helper or constructor you use to build a minimal valid `Node` instance suitable for `CaptureValue::Node` and `CaptureValue::Nodes`. If no helper exists, construct `Node` directly using its fields.
3. If your `captures` map uses a different value type (e.g. `Capture` wrapping `CaptureValue`), update the `captures` construction and the match arms accordingly:
   - Insert the appropriate wrapper into the `BTreeMap` values.
   - Adjust the `match` statements to destructure through that wrapper to reach the inner `CaptureValue`.
4. For a stronger equality check, if `CaptureValue` (and its payload types, including `Node`) can derive or implement `PartialEq`/`Eq`, you can replace the final `match`-based checks with direct equality assertions such as:
   - `assert_eq!(deserialized_captures.get("$node"), Some(&node_capture));`
   - `assert_eq!(deserialized_captures.get("$nodes"), Some(&nodes_capture));`
   This will verify that the entire enum variant and payload round-trip through serde JSON.
</issue_to_address>

### Comment 2
<location path="crates/sempai-core/src/test_support.rs" line_range="23-24" />
<code_context>
+}
+
+/// Error returned when a string does not match any known language name.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LanguageParseError {
+    name: String,
</code_context>
<issue_to_address>
**suggestion (testing):** Consider adding unit tests for QuotedString parsing success and failure cases

Since `QuotedString` is a core helper for the BDD step definitions, it would be valuable to add focused unit tests that:
- parse a valid input like `"\"foo\""` and verify `as_str()` returns `"foo"`, and
- cover malformed inputs (missing quotes, single quote, etc.), asserting a `QuotedStringParseError` is returned and that it includes the original input.
These tests would catch regressions in the quoting logic independently of the feature scenarios.
</issue_to_address>

### Comment 3
<location path="docs/execplans/5-1-1-entry-help.md" line_range="192" />
<code_context>
-  `fluent_and_fallback_outputs_are_identical` test guards against drift
-  between the two. Date/Author: 2026-02-26, updated 2026-02-28
+  Fluent messages, so translators can work on each line independently. The
+  English fallback values are centralized in a `bare_help` constants module in
+  `localizer.rs` so each string appears exactly once in the Rust source; the
+  `.ftl` file remains the canonical Fluent resource. The
</code_context>
<issue_to_address>
**issue (review_instructions):** Use en-GB spelling in documentation; replace the US spelling "centralized" with "centralised".

Use en-GB-oxendict spelling in this sentence. Replace "centralized" with "centralised" to comply with the documented style guide.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
Comments and documentation must use en-GB-oxendict spelling ("-ize" / "-yse" / "-our").

</details>
</issue_to_address>

### Comment 4
<location path="docs/users-guide.md" line_range="875" />
<code_context>
+The `Engine` struct exposes three methods for query compilation and execution:
+
+- `compile_yaml(yaml)` — compiles a YAML rule file into query plans.
+- `compile_dsl(rule_id, language, dsl)` — compiles a one-liner DSL expression.
+- `execute(plan, uri, source)` — executes a compiled plan against a source
+  snapshot.
</code_context>
<issue_to_address>
**issue (review_instructions):** The acronym “DSL” is used here without being expanded, which conflicts with the requirement to define uncommon acronyms on first use in this document.

To align with the style guidance, expand DSL the first time it appears in this file. For example:

- `compile_dsl(rule_id, language, dsl)` — compiles a one-liner domain-specific language (DSL) expression.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 5
<location path="docs/execplans/4-1-1-scaffold-sempai-core-and-sempai.md" line_range="203" />
<code_context>
+The design document at `docs/sempai-query-language-design.md` specifies six
+Sempai crates (Table 1, lines 67-74). This task scaffolds only the first two:
+
+- `crates/sempai-core` (`sempai_core`): Data model, diagnostics, planning IR.
+- `crates/sempai` (`sempai`): Facade crate, stable API. Re-exports from
+  `sempai_core` and provides the `Engine` struct.
</code_context>
<issue_to_address>
**issue (review_instructions):** The acronym “IR” is introduced here without expansion, which violates the requirement to define uncommon acronyms on first use.

Please expand IR the first time it appears. For example:

- `crates/sempai-core` (`sempai_core`): Data model, diagnostics, planning intermediate representation (IR).

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~
